File: android-ui/choose-currency/src/androidMain/kotlin/io/github/viacheslav/chugunov/moniqo/ui/choosecurrency/screen/ChooseCurrencyViewModel.kt
package io.github.viacheslav.chugunov.moniqo.ui.choosecurrency.screen

import androidx.lifecycle.viewModelScope
import io.github.viacheslav.chugunov.moniqo.core.model.Currency
import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyFilter
import io.github.viacheslav.chugunov.moniqo.core.model.CurrencyRates
import io.github.viacheslav.chugunov.moniqo.core.model.Rate
import io.github.viacheslav.chugunov.moniqo.core.usecase.GetCurrencyRatesUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.GetRatePairFlowUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.SaveFromRateUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.SaveToRateUseCase
import io.github.viacheslav.chugunov.moniqo.core.usecase.SetBaseRatesCurrencyUseCase
import io.github.viacheslav.chugunov.moniqo.ui.core.AppViewModel
import io.github.viacheslav.chugunov.moniqo.ui.core.navigation.CurrencySlot
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

internal class ChooseCurrencyViewModel(
    private val getCurrencyRatesUseCase: GetCurrencyRatesUseCase,
    private val getRatePairFlowUseCase: GetRatePairFlowUseCase,
    private val saveFromRateUseCase: SaveFromRateUseCase,
    private val saveToRateUseCase: SaveToRateUseCase,
    private val setRatesBaseCurrencyUseCase: SetBaseRatesCurrencyUseCase,
    private val mapper: ChooseCurrencyMapper,
) : AppViewModel<ChooseCurrencyState, ChooseCurrencyIntent, ChooseCurrencyEffect>(ChooseCurrencyState.Loading) {
    private var currencyRates: CurrencyRates? = null

    init {
        viewModelScope.launch {
            val rates = getCurrencyRatesUseCase()
            val pair = getRatePairFlowUseCase().first()
            currencyRates = rates
            updateState { mapper.toContent(rates, pair) }
        }
    }

    override fun onIntent(intent: ChooseCurrencyIntent) {
        when (intent) {
            is ChooseCurrencyIntent.Search -> handleSearch(intent.query)
            is ChooseCurrencyIntent.Filter -> handleFilter(intent.filter)
            is ChooseCurrencyIntent.SelectCurrency -> handleSelectCurrency(intent.currency, intent.slot)
        }
    }

    private fun handleSearch(query: String) {
        updateState { if (it is ChooseCurrencyState.Content) it.copy(query = query) else it }
    }

    private fun handleFilter(filter: CurrencyFilter) {
        updateState { if (it is ChooseCurrencyState.Content) it.copy(filter = filter) else it }
    }

    private fun handleSelectCurrency(
        currency: io.github.viacheslav.chugunov.moniqo.core.model.CurrencyInfo,
        slot: CurrencySlot,
    ) {
        viewModelScope.launch {
            val rates = currencyRates
            when (slot) {
                CurrencySlot.FROM -> {
                    val rate = rates?.findRate(currency.code) ?: return@launch
                    saveFromRateUseCase(rate)
                }
                CurrencySlot.TO -> {
                    val rate = rates?.findRate(currency.code) ?: return@launch
                    saveToRateUseCase(rate)
                }
                CurrencySlot.BASE -> {
                    viewModelScope.launch {
                        setRatesBaseCurrencyUseCase(Currency.of(currency.code))
                    }
                }
            }
            updateState { if (it is ChooseCurrencyState.Content) it.copy(query = "") else it }
            sendEffect(ChooseCurrencyEffect.NavigateBack)
        }
    }

    private fun CurrencyRates.findRate(code: String): Rate? {
        if (baseCurrency.name.uppercase() == code) {
            return Rate(baseCurrency, 1.0)
        }
        return rates.find { it.currency.name.uppercase() == code }
    }
}

File: iosApp/iosApp/choose-currency/screen/ChooseCurrencyViewModel.swift
import Foundation
import Shared

final class ChooseCurrencyViewModel: ObservableObject {
    @Published private(set) var state: ChooseCurrencyState = .loading
    @Published private(set) var shouldDismiss: Bool = false

    let slot: CurrencySlot
    private let mapper: ChooseCurrencyMapper

    private var currencyRates: CurrencyRates?
    private var recentCodes: [String] = []

    private var stopRatesLoad: (() -> Void)?
    private var stopRecentObservation: (() -> Void)?
    private var pendingJobs: [() -> Void] = []

    init(slot: CurrencySlot, mapper: ChooseCurrencyMapper) {
        self.slot = slot
        self.mapper = mapper
        startLoadingRates()
        startObservingRecent()
    }

    deinit {
        stopRatesLoad?()
        stopRecentObservation?()
        pendingJobs.forEach { $0() }
    }

    func onIntent(_ intent: ChooseCurrencyIntent) {
        switch intent {
        case .search(let query):
            updateContent { $0.query = query }
        case .filter(let filter):
            updateContent { $0.filter = filter }
        case .selectCurrency(let currency):
            handleSelectCurrency(currency)
        }
    }

    private func startLoadingRates() {
        stopRatesLoad = ChooseCurrencyBridgeKt.loadCurrencyRates { [weak self] rates in
            guard let self else { return }
            self.currencyRates = rates
            self.rebuildContent()
        }
    }

    private func startObservingRecent() {
        stopRecentObservation = ChooseCurrencyBridgeKt.observeRecentCurrencies { [weak self] codes in
            guard let self else { return }
            self.recentCodes = codes
            self.rebuildContent()
        }
    }

    private func rebuildContent() {
        guard let rates = currencyRates else { return }
        let preserved = currentContent
        var content = mapper.toContent(rates: rates, recentCodes: recentCodes)
        if let preserved {
            content.query = preserved.query
            content.filter = preserved.filter
        }
        state = .content(content)
    }

    private var currentContent: ChooseCurrencyContent? {
        if case .content(let c) = state { return c }
        return nil
    }

    private func updateContent(_ update: (inout ChooseCurrencyContent) -> Void) {
        guard case .content(var content) = state else { return }
        update(&content)
        state = .content(content)
    }

    private func handleSelectCurrency(_ currency: CurrencyInfo) {
        guard let rates = currencyRates else { return }
        switch slot {
        case .from:
            pendingJobs.append(ChooseCurrencyBridgeKt.saveFromRateByCode(rates: rates, code: currency.code))
        case .to:
            pendingJobs.append(ChooseCurrencyBridgeKt.saveToRateByCode(rates: rates, code: currency.code))
        case .base:
            pendingJobs.append(ChooseCurrencyBridgeKt.setBaseCurrencyByCode(rates: rates, code: currency.code))
        }
        pendingJobs.append(ChooseCurrencyBridgeKt.addRecentCurrency(code: currency.code))
        shouldDismiss = true
    }
}
